### ISSUE TITLE
Spring Circuitbreaker fallback not working with JDK (interface based) proxies 

### ISSUE THREAD

**Issue by emilevdh97** [resilience4j/resilience4j#2096]
**Resilience4j version: 2.2.0/LATEST
Java version: 17
Spring-boot: 3.2.1**

The `FallbackExecutor.java` class throws `java.lang.IllegalArgumentException: object is not an instance of declaring class`  while attempting to trigger the fallback function on spring beans defined with Interface-based proxies(JDK proxies) and using method level `@CircuitBreaker()` annotations. 

Minimal reproducible code: https://github.com/emilevdh97/R4JBugReplication

Current working version: 2.0.2
Suspected breaking Commit: https://github.com/resilience4j/resilience4j/commit/dc8d1559e261d5e586ce106d3d8a4269ff69c8fc

**Comment by RobWin** [comment]
Thank you.
Have you been able to debug it?
There seems to be a difference between using cglib or JDK proxies. 
Why do you plan to use JDK proxies? I think Spring Boot sets the default proxy to cglib.

**Comment by emilevdh97** [comment]
Yes, I've been able to debug the issue, the problem is caused by how the fallback method determines the target when using JDK Proxies, the java `Method.invoke` throws the exception because the proxy object used as the target is not of the correct type.
https://github.com/resilience4j/resilience4j/blob/24798b3aa800be28e28207da33f717583179be82/resilience4j-spring/src/main/java/io/github/resilience4j/fallback/FallbackMethod.java#L233

In R4J version 2.0.2, this was not an issue because there were no proxies involved at all. https://github.com/resilience4j/resilience4j/blob/0d3bb283a190307e2f9b99d80832f9471aa65da7/resilience4j-spring/src/main/java/io/github/resilience4j/fallback/FallbackMethod.java#L212

Based on the spring documentation for version 6.1.3, it mentions that ` Spring AOP defaults to using standard JDK dynamic proxies for AOP proxies`(https://docs.spring.io/spring-framework/reference/core/aop/introduction-proxies.html) but based on my experience, Spring generally defaults to a CGlib based proxy. 

The reason why we use JDK proxies is because we use Spring's WebServiceGatewaySupport to consume soap services and it doesn't work with CGlib proxies, To get the circuitbreakers working on it we had to use JDK proxies.

**Comment by dowenliu-xyz** [comment]
I've make a PR to try solve this problem.

**Comment by dowenliu-xyz** [comment]
I am the author of the IDEA plugin [Resilience4j Annotation Support](https://plugins.jetbrains.com/plugin/24265-resilience4j-annotation-support).

I found that this bug also affects when using annotations in the feign client. After that I tested MyBatis mapper interface, Spring data JDBC repository interface (they are both based on jdk proxy bean).

I have update my plugin to have a check at this issue:

https://github.com/resilience4j/resilience4j/assets/7844338/e7906a39-c6ba-4468-acea-2b9200f06048

**Comment by dowenliu-xyz** [comment]
@emilevdh97  Hi, I've found a new workaround:
If you define a default method for `fallbackMethod` in the interface (it doesn't matter what is done in the `default` method), just override the default method in the implementing class and the `fallbackMethod` will work.

I've been experimenting with this in my plays project: 

https://github.com/dowenliu-xyz/resilience4j-plays/blob/6591cf6b26d23c5d22eeab195c59cd44d75ecb45/play05-jdk-proxy/src/main/java/org/example/demo/Runner.java#L134
```Java
            expectProxyType(annoBaseFallbackDefaultOverriddenJavaInterfaceForJava, ProxyType.JDK);
            expectFallbackTakeEffect(annoBaseFallbackDefaultOverriddenJavaInterfaceForJava::greeting);
```
https://github.com/dowenliu-xyz/resilience4j-plays/blob/main/play05-jdk-proxy/src/main/java/org/example/demo/demos/annobase/interfaces/AnnoBaseFallbackDefaultOverriddenJavaInterfaceForJava.java
```Java
public interface AnnoBaseFallbackDefaultOverriddenJavaInterfaceForJava {
    @NotNull
    String greeting(@Nullable String name);

    @NotNull
    default String fallback(@Nullable String name, @Nullable Throwable thr) {
        return "";
    }
}
```
https://github.com/dowenliu-xyz/resilience4j-plays/blob/main/play05-jdk-proxy/src/main/java/org/example/demo/demos/annobase/interfaces/AnnoBaseFallbackDefaultOverriddenJavaInterfaceJavaDemo.java
```Java
public class AnnoBaseFallbackDefaultOverriddenJavaInterfaceJavaDemo implements AnnoBaseFallbackDefaultOverriddenJavaInterfaceForJava {
    @CircuitBreaker(name = "demo", fallbackMethod = "fallback")
    @NotNull
    @Override
    public String greeting(@Nullable String name) {
        return doGreeting(name);
    }

    @NotNull
    @Override
    public String fallback(@Nullable String name, @Nullable Throwable thr) {
        return doFallback(name, thr);
    }
}
```
